go-controller/pkg/ovn/obj_retry.go

```diff
@@ -126,18 +189,1067 @@ func (r *retryObjs) getObjRetryEntry(key string) *retryObjEntry {
 	return nil
 }
 
-// triggerRetryObjs track the retry objects map and every 30 seconds check if any object need to be retried
-func (oc *Controller) triggerRetryObjs(retryChan chan struct{}, iterateObjs func(bool)) {
-	go func() {
-		for {
-			select {
-			case <-time.After(retryObjInterval):
-				iterateObjs(false)
-			case <-retryChan:
-				iterateObjs(true)
-			case <-oc.stopChan:
+var sep = "/"
+
+func splitNamespacedName(namespacedName string) (string, string) {
+	if strings.Contains(namespacedName, sep) {
+		s := strings.SplitN(namespacedName, sep, 2)
+		if len(s) == 2 {
+			return s[0], s[1]
+		}
+	}
+	return namespacedName, ""
+}
+
+func getNamespacedName(namespace, name string) string {
+	return namespace + sep + name
+}
+
+// hasResourceAnUpdateFunc returns true if the given resource type has a dedicated update function.
+// It returns false if, upon an update event on this resource type, we instead need to first delete the old
+// object and then add the new one.
+func hasResourceAnUpdateFunc(objType reflect.Type) bool {
+	switch objType {
+	case factory.PodType,
+		factory.NodeType,
+		factory.PeerPodSelectorTypeVar,
+		factory.PeerPodForNamespaceAndPodSelectorType,
+		factory.LocalPodSelectorTypeVar:
+		return true
+	}
+	return false
+}
+
+// areResourcesEqual returns true if, given two objects of a known resource type, the update logic for this resource
+// type considers them equal and therefore no update is needed. It returns false when the two objects are not considered
+// equal and an update needs be executed. This is regardless of how the update is carried out (whether with a dedicated update
+// function or with a delete on the old obj followed by an add on the new obj).
+func areResourcesEqual(objType reflect.Type, obj1, obj2 interface{}) (bool, error) {
+	// switch based on type
+	switch objType {
+	case factory.PolicyType:
+		NP1, ok := obj1.(*knet.NetworkPolicy)
+		if !ok {
+			return false, fmt.Errorf("could not cast obj1 of type interface{} to *knet.NetworkPolicy")
+		}
+		NP2, ok := obj2.(*knet.NetworkPolicy)
+		if !ok {
+			return false, fmt.Errorf("could not cast obj2 of type interface{}  to *knet.NetworkPolicy")
+		}
+		return reflect.DeepEqual(NP1, NP2), nil
+
+	case factory.NodeType:
+		node1, ok := obj1.(*kapi.Node)
+		if !ok {
+			return false, fmt.Errorf("could not cast obj1 of type interface{} to *kapi.Node")
+		}
+		node2, ok := obj2.(*kapi.Node)
+		if !ok {
+			return false, fmt.Errorf("could not cast obj2 of type interface{} to *kapi.Node")
+		}
+
+		// when shouldUpdate is false, the hostsubnet is not assigned by ovn-kubernetes
+		shouldUpdate, err := shouldUpdate(node2, node1)
+		if err != nil {
+			klog.Errorf(err.Error())
+		}
+		return !shouldUpdate, nil
+
+	case factory.PeerServiceTypeVar:
+		service1, ok := obj1.(*kapi.Service)
```

Comment:
does it make sense that this function implement shouldUpdate(obj1, obj2) and let each resource implement its own ?
the idea from this file to keep it generic if we can check everytype and do resource specific then I would say it shouldn't belong to this file ?

Reply:
you mean I should move `shouldUpdate` to this file? Or something else? :)